### PR TITLE
feat: add npm engine constraint properties

### LINKED ISSUE #1506: feat: `Contrib.FromNodePackageJsonBuilders.Builders.ComponentBuilder` detail runtime dependencies
based on https://github.com/CycloneDX/cyclonedx-node-yarn/issues/448

----

## Is your feature request related to a problem? Please describe.

i am detailing my runtime to`node` via `package.json::engines` - see <https://docs.npmjs.com/cli/v11/configuring-npm/package-json#engines>
```json
{
  "engines": {
    "node": ">=0.10.3 <15"
  }
}
```

I want this information being detailed in the Component generated by  `Contrib.FromNodePackageJsonBuilders.Builders.ComponentBuilder`.

## Describe the solution you'd like

the sourced information may stem from package manifest (`package.json`) or from lockfile or from `npm-ls`

* populate `component.properties` with `cdx:npm:package:constraint:engine:<name> = <version range>` being whatever value found in the package manifest. as per <https://cyclonedx.github.io/cyclonedx-property-taxonomy/cdx/npm.html>
* optional, for known JS runtimes: for CycloneDX 1.7 add a component to the components list, that has the following attributes:
  * type: `platform` for `node`, `deno`, etc ... 
  * group: empty
  * name: `node`, or `deno` or whatever it is ...
  * version: unset
  * versionRange: a VERS value `vers:npm/...`  - see https://github.com/package-url/vers-spec/blob/main/VERSION-RANGE-SPEC.md
  * isExternal: `true`
  * packageUrl: empty, as we dont know where this external component shall be taken from
  * have this new component a dependency of the other component

## Describe alternatives you've considered

none

## Additional context

docs: 
- packageJSON "engines": <https://docs.npmjs.com/cli/v11/configuring-npm/package-json#engines>
- CDX properties taxonomy `cdx:npm:` <https://cyclonedx.github.io/cyclonedx-property-taxonomy/cdx/npm.html>
- CDX "isExternal" feature: <https://github.com/CycloneDX/specification/pull/586>
- CDX "versionRange" feature: <https://github.com/CycloneDX/specification/pull/647>
- VERS spec: <https://github.com/package-url/vers-spec/blob/main/VERSION-RANGE-SPEC.md>

## Contribution 

- [ ] I am willing to provide an implementation
- [x] I will wait until somebody else implements it

### PR DESCRIPTION
**Pull request by CAOShurong**
### Description

Populate `Component.properties` from the string-valued entries in a Node
package manifest's `engines` object. Each entry uses the existing CycloneDX
npm property taxonomy name
`cdx:npm:package:constraint:engine:<name>`.

Malformed non-string engine constraints are ignored, consistent with the
builder's defensive handling of other package manifest fields.

Fixes issue: #1506

### Verification

- `npm run test:node` (4,166 passing)
- `npm run test:lint`
- `npm run test:standard`
- `npm run test:dependencies`
- `npm run build:node`
- `npm run build:web`
- `npx tsc -b ./tsconfig.d.json`
- `npm pack --ignore-scripts`, followed by installation into a clean directory
  and runtime checks for valid constraints, malformed values, and a missing
  package name

### AI Tool Disclosure

- [ ] My contribution does not include any AI-generated content
- [x] My contribution includes AI-generated content, as disclosed below:
  - AI Tools: `OpenAI Codex`
  - LLMs and versions: `GPT-5`
  - Prompts: `Inspect issue #1506 and the repository contribution rules; implement only the required npm engine constraint properties; add positive and malformed-input regression coverage; run build, lint, full tests, package, and clean-install runtime verification.`

### Affirmation

- [x] My code follows the [CONTRIBUTING.md](https://github.com/CycloneDX/cyclonedx-javascript-library/blob/main/CONTRIBUTING.md) guidelines

### CODE REVIEW

File: src/contrib/fromNodePackageJson/builders.ts [hunk 1]
```diff
@@ -30,12 +30,28 @@ import { ComponentType } from '../../enums/componentType'
...
 import { ExternalReferenceRepository } from '../../models/externalReference'
 import { LicenseRepository } from '../../models/license'
+import { Property, PropertyRepository } from '../../models/property'
 import { Tool } from '../../models/tool'
 import type { LicenseFactory } from '../license/factories'
 import { splitNameGroup } from './_helpers/packageJson'
 import type { ExternalReferenceFactory } from './factories'
 import type { NodePackageJson } from './types'
 
+function makeEngineProperties (engines: unknown): PropertyRepository {
```
Comment (jkowalleck) [inline src/contrib/fromNodePackageJson/builders.ts]: please move this inside the `ComponentBuilder`

File: src/contrib/fromNodePackageJson/builders.ts [hunk 2]
```diff
@@ -144,6 +162,7 @@ export class ComponentBuilder {
       externalReferences: new ExternalReferenceRepository(externalReferences),
       group,
       licenses,
+      properties,
```
Comment (jkowalleck) [inline src/contrib/fromNodePackageJson/builders.ts]: `properties` at the bottom

File: tests/integration/Contrib.FromNodePackageJson.Builders.ComponentBuilder.node.test.js [hunk 3]
```diff
@@ -84,10 +88,25 @@ suite('integration: Contrib.FromNodePackageJson.Builders.ComponentBuilder', () =
...
+          properties: new Models.PropertyRepository([
+            new Models.Property('cdx:npm:package:constraint:engine:node', '>=20.18.0'),
+            new Models.Property('cdx:npm:package:constraint:engine:npm', '^10.8.2'),
+          ]),
           version: `1.33.7-alpha.23.${salt}`
         }
       )
     ],
+    [
+      'ignore malformed engine constraints',
```
Comment (jkowalleck) [inline tests/integration/Contrib.FromNodePackageJson.Builders.ComponentBuilder.node.test.js:100]: pleasse also adda test case where the malformed `engines` is an array, and one for where it is `null`, and one where it is a string.

File: HISTORY.md [hunk 4]
```diff
@@ -6,6 +6,11 @@ All notable changes to this project will be documented in this file.
 
 <!-- add unreleased items here -->
 
+* Added
+  * Populate npm engine constraint properties from `package.json` ([#1506])
+
+[#1506]: https://github.com/CycloneDX/cyclonedx-javascript-library/issues/1506
+
```
Comment (jkowalleck) [inline HISTORY.md:14]:
```suggestion
[#1507]: https://github.com/CycloneDX/cyclonedx-javascript-library/pull/1507

```

File: HISTORY.md [hunk 5]
```diff
@@ -6,6 +6,11 @@ All notable changes to this project will be documented in this file.
 
 <!-- add unreleased items here -->
 
+* Added
+  * Populate npm engine constraint properties from `package.json` ([#1506])
```
Comment (jkowalleck) [inline HISTORY.md]:
```suggestion
  * `Contrib.FromNodePackageJson.Builders.ComponentBuilder` may populate properties with engine constraint ([#1506] via [#1507])
```

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by CAOShurong** [comment]
Both review requests are now in place at exact head `dee02ea`:

- `makeEngineProperties` moved inside `ComponentBuilder` and the generated `properties` are passed at the bottom of the component options.
- Added regression cases where malformed `engines` is an array, `null` and a string (plus the existing non-string-constraint case).

Verification on this head: 8 focused ComponentBuilder tests, 4,155 full node tests, `tsc` noEmit, eslint, knip dependency checks, node/web/declaration builds, and a real installed-dist probe covering valid, array, null, string and missing `engines` all pass. CI is now running on the updated head.

Thanks for the review and for applying the HISTORY.md suggestion.

**Review by jkowalleck — APPROVED** [review]

**Review by jkowalleck — APPROVED** [review]

**Comment by jkowalleck** [comment]
released via https://github.com/CycloneDX/cyclonedx-javascript-library/releases/tag/v10.2.0
